extensions/amp-audio/1.0/component.js

```diff
@@ -0,0 +1,195 @@
+
+const {useCallback, useEffect, useImperativeHandle, useMemo, useRef} = Preact;
+
+/**
+ * @param {!AudioDef.Props} props
+ * @param {{current: (!AudioDef.AudioApi|null)}} ref
+ * @return {PreactDef.Renderable}
+ */
+export function AudioWithRef(props, ref) {
+  const audioRef = useRef();
+
+  const {
+    album,
+    artist,
+    artwork,
+    autoplay,
+    children,
+    controlsList,
+    isStoryDescendant_,
+    loop,
+    muted,
+    preload,
+    propagateAttributes,
+    src,
+    title,
+    toggleFallback,
+    validateMediaMetadata,
+    ...rest
+  } = props;
+
+  /** @public {boolean} */
+  const isPlaying = useRef(false);
+
+  /**
+   * Checks if the function is allowed to be called
+   * @return {boolean}
+   */
+  const isInvocationValid_ = useCallback(() => {
+    if (!audioRef) {
+      return false;
+    }
+
+    if (isStoryDescendant_ && isStoryDescendant_()) {
+      console /*OK*/
+        .warn(
+          '<amp-story> elements do not support actions on <amp-audio> elements'
+        );
+      return false;
+    }
+    return true;
+  }, [isStoryDescendant_]);
+
+  /**
+   * Prepares Media Metadata
+   */
+  const metaData = useMemo(() => {
+    return {
+      title,
+      artist,
+      album,
+      artwork: [{src: artwork}],
+    };
+  }, [title, artist, album, artwork]);
+
+  /**
+   * Plays audio callback
+   */
+  const playCallback = useCallback(() => {
+    triggerAnalyticsEvent(audioRef.current, 'audio-play');
+
+    if (!isInvocationValid_()) {
+      return;
+    }
+    audioRef.current.play();
+    isPlaying.current = true;
+  }, [isPlaying, isInvocationValid_]);
+
+  /**
+   * Pauses audio callback
+   */
+  const pauseCallback = useCallback(() => {
+    triggerAnalyticsEvent(audioRef.current, 'audio-pause');
+
+    if (!isInvocationValid_()) {
+      return;
+    }
+    audioRef.current.pause();
+    isPlaying.current = false;
+  }, [isPlaying, isInvocationValid_]);
+
+  /**
+   * Updates media session for current window/tab
+   */
+  const audioPlaying = useCallback(() => {
+    const win = audioRef.current?.ownerDocument?.defaultView;
+    const element = audioRef.current;
+
+    if (validateMediaMetadata) {
+      validateMediaMetadata(element, metaData);
+    }
+
+    setMediaSession(win, metaData, playCallback, pauseCallback);
+  }, [metaData, validateMediaMetadata, playCallback, pauseCallback]);
+
+  useEffect(() => {
+    if (!audioRef.current.play) {
+      if (toggleFallback) {
+        toggleFallback(true);
+      }
+      return;
+    }
+
+    const unlistenPlaying = listen(audioRef.current, 'playing', () =>
+      audioPlaying()
+    );
+
+    if (propagateAttributes) {
+      propagateAttributes(
+        [
+          'aria-describedby',
+          'aria-label',
+          'aria-labelledby',
+          'autoplay',
+          'controlsList',
+          'loop',
+          'muted',
+          'preload',
+          'src',
+        ],
+        audioRef.current
+      );
+    }
```

Comment:
Same as above - is this needed? Sorry I am not too familiar with its purpose 😅

Reply:
I am also not sure about it 🤔

From the function information, I came to the conclusion that it only propagate attribute that are not `null`:
https://github.com/ampproject/amphtml/blob/3fed6703d6c0b2c1be641876f32c0e195398e3ee/src/core/dom/propagate-attributes.js#L19-L28

I copied logic from AMP version of `amp-audio`:
https://github.com/ampproject/amphtml/blob/3fed6703d6c0b2c1be641876f32c0e195398e3ee/extensions/amp-audio/0.1/amp-audio.js#L130-L144

Let me explore in depth. Any idea/suggestion would be helpful 💡